core/node/rpc/sync/client/remote.go

```diff
@@ -0,0 +1,257 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/ethereum/go-ethereum/common"
+	. "github.com/river-build/river/core/node/base"
+	"github.com/river-build/river/core/node/dlog"
+	. "github.com/river-build/river/core/node/protocol"
+	"github.com/river-build/river/core/node/protocol/protocolconnect"
+	. "github.com/river-build/river/core/node/shared"
+)
+
+type remoteSyncer struct {
+	syncStreamCtx    context.Context
+	syncStreamCancel context.CancelFunc
+	syncID           atomic.Value
+	forwarderSyncID  string
+	remoteAddr       common.Address
+	client           protocolconnect.StreamServiceClient
+	cookies          []*SyncCookie
+	messages         chan<- *SyncStreamsResponse
+	streams          sync.Map
+	responseStream   *connect.ServerStreamForClient[SyncStreamsResponse]
+}
+
+func newRemoteSyncer(
+	ctx context.Context,
+	forwarderSyncID string,
+	remoteAddr common.Address,
+	client protocolconnect.StreamServiceClient,
+	cookies []*SyncCookie,
+	messages chan<- *SyncStreamsResponse,
+) (*remoteSyncer, error) {
+	syncStreamCtx, syncStreamCancel := context.WithCancel(ctx)
+	responseStream, err := client.SyncStreams(syncStreamCtx, connect.NewRequest(&SyncStreamsRequest{SyncPos: cookies}))
+	if err != nil {
+		for _, cookie := range cookies {
+			messages <- &SyncStreamsResponse{
+				SyncOp:   SyncOp_SYNC_DOWN,
+				StreamId: cookie.GetStreamId(),
+			}
+		}
+		syncStreamCancel()
+		return nil, err
+	}
+
+	if !responseStream.Receive() {
+		syncStreamCancel()
+		return nil, responseStream.Err()
+	}
+
+	log := dlog.FromCtx(ctx)
+
+	if responseStream.Msg().SyncOp != SyncOp_SYNC_NEW || responseStream.Msg().SyncId == "" {
+		log.Error("Received unexpected sync stream message",
+			"syncOp", responseStream.Msg().SyncOp,
+			"syncId", responseStream.Msg().SyncId)
+		syncStreamCancel()
+		return nil, err
+	}
+
+	s := &remoteSyncer{
+		forwarderSyncID:  forwarderSyncID,
+		syncStreamCtx:    syncStreamCtx,
+		syncStreamCancel: syncStreamCancel,
+		client:           client,
+		cookies:          cookies,
+		messages:         messages,
+		responseStream:   responseStream,
+		remoteAddr:       remoteAddr,
+	}
+
+	s.syncID.Store(responseStream.Msg().SyncId)
+
+	for _, cookie := range s.cookies {
+		streamID, _ := StreamIdFromBytes(cookie.GetStreamId())
+		s.streams.Store(streamID, struct{}{})
+	}
+
+	return s, nil
+}
+
+func (s *remoteSyncer) Run() {
+	defer s.responseStream.Close()
+
+	var latestMsgReceived atomic.Value
+
+	latestMsgReceived.Store(time.Now())
+
+	go func() {
```

Comment:
please put this func in the method and go-call it here: it's relatively long and makes Run() less readable as such